### PR TITLE
Feature 53107 url compact

### PR DESCRIPTION
**Pull request by psunix**

### CODE REVIEW

File: app/controllers/accounts_controller.rb [hunk 1]
```diff
@@ -132,29 +113,39 @@ def move_account(format, target_group)
...
+    @account ||= Account.find(params[:id])
+  end
+
+  def accounts_breadcrumbs
+    teams_breadcrumbs
 
-    if action_name == 'show' || action_name == 'edit'
-      add_breadcrumb @group.label, :team_group_accounts_path
-      add_breadcrumb @account.label
+    if %w[show edit new].include?(action_name)
```
Comment (psunix) [inline app/controllers/accounts_controller.rb]: new ?
you might show breadcrumbs on new ... but then you have to use another label Line 132
something like "Account"

File: app/controllers/groups_controller.rb [hunk 2]
```diff
@@ -8,32 +8,20 @@
 class GroupsController < ApplicationController
   self.permitted_attrs = [:name, :description]
 
+  before_action :group, except: [:new, :create]
```
Comment (psunix) [inline app/controllers/groups_controller.rb]: is this needed ?
the idea of the group getter is to cache it as soon as it's required. 
so if you access group by group and not by @group, this should work perfectly

File: app/controllers/legacy_routes/handler.rb [hunk 3]
```diff
@@ -5,13 +5,13 @@
 #  See the COPYING file at the top-level directory or at
 #  https://github.com/puzzle/cryptopus.
 
-class LegacyRoutes::Locale
+class LegacyRoutes::Handler
```
Comment (psunix) [inline app/controllers/legacy_routes/handler.rb]: name this RedirectUrl instead of Handler

File: app/controllers/legacy_routes_controller.rb [hunk 4]
```diff
@@ -17,7 +17,6 @@ def redirect
   private
 
   def redirect_url
-    LegacyRoutes::Locale.new.redirect_url(request.path)
+    LegacyRoutes::Handler.new.redirect_url(request.path)
```
Comment (psunix) [inline app/controllers/legacy_routes_controller.rb]: LegacyRoutes::RedirectUrl.new(request.path).redirect_to

File: app/controllers/legacy_routes/handler.rb [hunk 5]
```diff
@@ -5,13 +5,13 @@
...
 
   LOCALES_REGEX = I18n.available_locales.map do |locale|
     "\/#{locale}|"
   end.join('').chop.freeze
 
   def redirect_url(url)
-    url.downcase.remove(/#{LOCALES_REGEX}/)
+    url.downcase.remove(/#{LOCALES_REGEX}/).sub(/(accounts|groups)$/, '\2')
```
Comment (psunix) [inline app/controllers/legacy_routes/handler.rb]: def redirect_to
   remove_locale
end

use ruby case ... https://www.rubyguides.com/2015/10/ruby-case/

when legacy_accounts_path?
   account_path(account_id)

\# ../teams/42/groups/33/accounts/45 -> /accounts/45 
def legacy_accounts_path?

add include Rails.application.routes.url_helpers to get all path/url helpers in controller

File: test/integration/create_account_test.rb [hunk 6]
```diff
@@ -4,7 +4,7 @@
 #  Cryptopus and licensed under the Affero General Public License version 3 or later.
 #  See the COPYING file at the top-level directory or at
 #  https://github.com/puzzle/cryptopus.
-
+require 'pry'
```
Comment (psunix) [inline test/integration/create_account_test.rb]: pry ... never check in pry ..

Comment (psunix) [reply test/integration/create_account_test.rb]: require 'pry'; binding.pry 
you can use it like this as a one-liner. no need to add it on top of a file

File: test/integration/legacy_routes/handler_test.rb [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+#  See the COPYING file at the top-level directory or at
+#  https://github.com/puzzle/cryptopus.
+require 'pry'
+require 'test_helper'
+
+class LegacyRoutes::UrlTest < ActionDispatch::IntegrationTest
+  include IntegrationTest::DefaultHelper
+
+  test 'redirects to groups with without groups in url' do
+    account_url = "/teams/#{teams(:team1).id}/groups/#{groups(:group1).id}"
```
Comment (psunix) [inline test/integration/legacy_routes/handler_test.rb]: this is not a account_url, is it ?
name this legacy_account_url ... this makes your test much more readable

File: test/integration/legacy_routes/handler_test.rb [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+#  Copyright (c) 2008-2017, Puzzle ITC GmbH. This file is part of
+#  Cryptopus and licensed under the Affero General Public License version 3 or later.
+#  See the COPYING file at the top-level directory or at
+#  https://github.com/puzzle/cryptopus.
+require 'pry'
+require 'test_helper'
+
+class LegacyRoutes::UrlTest < ActionDispatch::IntegrationTest
+  include IntegrationTest::DefaultHelper
+
```
Comment (psunix) [inline test/integration/legacy_routes/handler_test.rb]: why not add a comment above test to explain how this works ....
e.g. \# /de/teams/42/groups/43/accounts/42 -> /accounts/42
then you can use a pretty short name for the test case ... test 'legacy groups url'

a legacy route always includes the locale !

File: test/integration/legacy_routes/handler_test.rb [hunk 9]
```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+#  Copyright (c) 2008-2017, Puzzle ITC GmbH. This file is part of
+#  Cryptopus and licensed under the Affero General Public License version 3 or later.
+#  See the COPYING file at the top-level directory or at
+#  https://github.com/puzzle/cryptopus.
+require 'pry'
```
Comment (psunix) [inline test/integration/legacy_routes/handler_test.rb]: no pry ...

File: test/integration/legacy_routes/handler_test.rb [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+require 'test_helper'
+
+class LegacyRoutes::UrlTest < ActionDispatch::IntegrationTest
```
Comment (psunix) [inline test/integration/legacy_routes/handler_test.rb]: Integration Tests do not have to be named like e.g. controllers. 

LegacyRoutesRedirectTest

File: test/integration/legacy_routes/handler_test.rb [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    login_as('bob')
+
+    get account_url
+
+    assert_redirected_to redirect_url
+  end
+
+  test 'redirects to groups url with groups in uppercase in url' do
+    redirect_url = "/teams/#{teams(:team1).id}/groups/#{groups(:group1).id}/"
+    accounts_url = "/teams/#{teams(:team1).id}/groups/#{groups(:group1).id}/ACCOUNTS/"
```
Comment (psunix) [inline test/integration/legacy_routes/handler_test.rb]: we don't have to check for UPERCASE urls !!!!!!

File: test/integration/legacy_routes/handler_test.rb [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    assert_redirected_to redirect_url
+  end
+
+  test 'redirects to groups with groups in url' do
+    team_url = "/teams/#{teams(:team1).id}/groups"
+    login_as('bob')
+
+    get team_url
+
+    assert_routing "/teams/#{teams(:team1).id}", controller: "teams", action: "show", id: "#{teams(:team1).id}"
```
Comment (psunix) [inline test/integration/legacy_routes/handler_test.rb]: why not assert_redirected_to ?

File: app/controllers/legacy_routes/handler.rb (hunk 3 above)
Comment (psunix) [reply app/controllers/legacy_routes/handler.rb]: there's no unit test for this class ?!
TDD ... make sure you handle all urls in this unit test first

File: app/controllers/items_controller.rb [hunk 13]
```diff
@@ -44,23 +45,30 @@ def destroy
...
     end
   end
 
   private
 
-  def load_parents
-    @group = team.groups.find(params[:group_id])
-    @account = @group.accounts.find(params[:account_id])
+  def account
+    @account = Account.find(params[:account_id])
```
Comment (Thunderguts13) [inline app/controllers/items_controller.rb:55]: @account ||= Account.find

File: app/controllers/items_controller.rb [hunk 14]
```diff
@@ -8,7 +8,8 @@
 class ItemsController < ApplicationController
   self.permitted_attrs = [:description, :file]
 
-  before_action :load_parents
+  before_action :account
+  before_action :group
```
Comment (Thunderguts13) [inline app/controllers/items_controller.rb:12]: Remove the before actions and use `account` and not `@account` same for group

File: app/controllers/teams_controller.rb [hunk 15]
```diff
@@ -20,6 +20,20 @@ def index
...
   end
 
+  # GET /teams/1
+  def show
+    authorize team, :team_member?
+    @groups = team.groups
+    groups_breadcrumbs
+
+    teammembers = team.teammembers.list
+    @teammembers = teammembers.includes(:user).sort_by { |tm| tm.label.downcase }
```
Comment (Thunderguts13) [inline app/controllers/teams_controller.rb:30]: Is this needed?

File: config/routes.rb [hunk 16]
```diff
@@ -16,38 +16,17 @@
...
-
-  namespace :admin do
-    resources :maintenance_tasks, only: :index
-    post '/maintenance_tasks/:id/execute', to: 'maintenance_tasks#execute', as: 'maintenance_tasks_execute'
-
-    resource :settings do
-      post 'update_all'
-      get 'index'
-    end
+    resources :groups, except: [:index] do
```
Comment (Thunderguts13) [inline config/routes.rb:22]: remove the do/end

File: spec/integration/legacy_routes/redirect_url.rb [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    login_as('bob')
+
+    get legacy_account_url
+
+    assert_redirected_to team_group_path(teams(:team1).id, groups(:group1).id)
+  end
+
+  # /de/teams/1/groups/1/accounts -> /teams/1/groups/1/
+  it 'redirects to accounts url with accounts in legacy url' do
+    legacy_account_url = "/de/teams/#{teams(:team1).id}/groups/#{groups(:group1).id}/accounts"
```
Comment (Thunderguts13) [inline spec/integration/legacy_routes/redirect_url.rb:20]: use some other languages too

File: spec/integration/legacy_routes/redirect_url.rb [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    assert_redirected_to search_path
+  end
+
+  # /de/admin/settings/index -> /admin/settings/index
+  it 'redirects to admin settings url without locale' do
+    login_as('bob')
+    get '/de/admin/settings/index'
+
+    assert_redirected_to admin_settings_path
+  end
```
Comment (Thunderguts13) [inline spec/integration/legacy_routes/redirect_url.rb:108]: make a test for an invalid route

File: app/controllers/groups_controller.rb [hunk 19]
```diff
@@ -10,27 +10,14 @@ class GroupsController < ApplicationController
...
-    respond_to do |format|
-      format.html # index.html.haml
-    end
-  end
-
   # GET /teams/1/groups/1
   def show
-    @group = team.groups.find(params[:id])
-    authorize @group
+    authorize team
```
Comment (psunix) [inline app/controllers/groups_controller.rb:15]: well, the result is the same in this case for the moment ... but we should not authorize on team here !

File: app/controllers/groups_controller.rb [hunk 20]
```diff
@@ -10,27 +10,14 @@ class GroupsController < ApplicationController
...
+    @accounts = group.accounts
+    groups_breadcrumbs
 
     respond_to do |format|
-      format.html # show.html.haml
+      format.html # index.html.haml
```
Comment (psunix) [inline app/controllers/groups_controller.rb:20]: just a comment ... but index.html.haml does not exist !

File: app/controllers/items_controller.rb [hunk 21]
```diff
@@ -8,12 +8,11 @@
 class ItemsController < ApplicationController
   self.permitted_attrs = [:description, :file]
 
-  before_action :load_parents
   helper_method :team
 
   # GET /teams/1/groups/1/accounts/new
```
Comment (psunix) [inline app/controllers/items_controller.rb:13]: incorrect url !

File: app/controllers/legacy_routes/redirect_url.rb [hunk 22]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
```
Comment (psunix) [inline app/controllers/legacy_routes/redirect_url.rb:1]: nice! well done !

File: config/locales/ru.yml [hunk 23]
```diff
@@ -56,9 +56,6 @@ ru:
```
Comment (psunix) [inline config/locales/ru.yml:56]: nice ... all locale files cleaned up !

File: config/routes.rb [hunk 24]
```diff
@@ -16,38 +16,15 @@
...
+    get "/:id/groups" => "legacy_routes#redirect"
     resources :api_users, only: [:index, :create, :destroy], module: 'api/team'
     resources :teammembers
-    resources :groups do
-      resources :accounts do
-        put 'move', to: 'accounts#move'
-        resources :items
-      end
-    end
+    resources :groups, except: [:index]
```
Comment (psunix) [inline config/routes.rb:22]: moved this to line 19 ... so line 19 with redirect get obsolete

File: config/routes.rb [hunk 25]
```diff
@@ -16,38 +16,15 @@
   end
 
   resources :teams do
+    get "/:id/groups" => "legacy_routes#redirect"
```
Comment (psunix) [inline config/routes.rb:19]: removed ....

File: spec/integration/legacy_routes/redirect_url_spec.rb [hunk 26]
```diff
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe LegacyRoutes::RedirectUrl do
+  include IntegrationTest::DefaultHelper
+
+  # /de/teams/1/groups/1 -> /teams/1/groups/1/
+  it 'redirects to groups without groups in url' do
```
Comment (psunix) [inline spec/integration/legacy_routes/redirect_url_spec.rb:9]: very nice !

File: spec/integration/legacy_routes/redirect_url_spec.rb [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+    team1 = teams(:team1)
+    group1 = groups(:group1)
+
+    invalid_account_url = "/teams/#{team1.id}/groups/#{group1.id}/accountes"
+    login_as('bob')
+
+    expect { get invalid_account_url }.to raise_error(ActionController::RoutingError)
+  end
+
+  # /ch_vs/teams/1/groups/1/accounts -> RoutingError
```
Comment (psunix) [inline spec/integration/legacy_routes/redirect_url_spec.rb:156]: +1 for ch_vs :joy:

**Comment by psunix** [comment]
created a new branch and merged it.

**Pull request closed without merging**